### PR TITLE
refactor(hydro_deploy)!: use `buildstructor` to handle excessive `Deployment` method arguments, fix #1364

### LINKED ISSUE #1364: refactor(hydro_deploy): use a builder for `GcpComputeEngineHost` and `AzureHost`
Yeah, we should replace this with a builder like for `HydroflowCrate`.

_Originally posted by @shadaj in https://github.com/hydro-project/hydroflow/pull/1313#discussion_r1694006732_

### PR DESCRIPTION
**Pull request by MingweiSamuel**
Adds new method `Deployment::AzureHost`

BREAKING CHANGE: Changes syntax for `Deployment::GcpComputeEngineHost`

### CODE REVIEW

File: hydro_deploy/core/src/deployment.rs [hunk 1]
```diff
@@ -186,3 +162,47 @@ impl Deployment {
         arc
     }
 }
+
+/// Buildstructor methods.
+#[buildstructor::buildstructor]
+impl Deployment {
+    #[allow(clippy::too_many_arguments)]
+    #[builder(entry = "GcpComputeEngineHost", exit = "add")]
```
Comment (shadaj) [inline hydro_deploy/core/src/deployment.rs:170]: I wonder if it is possible to have `exit` be just `drop` (or is that asking for too much haha)?

Comment (MingweiSamuel) [reply hydro_deploy/core/src/deployment.rs:170]: Hmm it is theoretically possible but the macro isn't implemented to support it -- also I think that would be _too_ fancy. Though the macro ought to mark the struct #[must_use] https://github.com/BrynCooke/buildstructor/issues/189

Comment (MingweiSamuel) [reply hydro_deploy/core/src/deployment.rs:170]: Perhaps should be called `GcpComputeEngineHostBuilder` and `.build()`

**Review by shadaj — APPROVED** [review]
Way nicer than before!

**Pull request merged**